monitoring/bin/change_grafana_admin_password.sh

```diff
@@ -0,0 +1,133 @@
+        show_usage
+        exit 2
+      fi
+      ;;
+    -p|--password)
+      if [ -n "$2" ] && [ ${2:0:1} != "-" ]; then
+        password=$2
+        shift 2
+      else
+        log_error "A value for parameter [PASSWORD] has not been provided." >&2
+        show_usage
+        exit 2
+      fi
+      ;;
+    -h|--help)
+      show_usage
+      exit
+      ;;
+    -*|--*=) # unsupported flags
+      log_error "Unsupported flag $1" >&2
+      show_usage
+      exit 1
+      ;;
+    *) # preserve positional arguments
+      POS_PARMS="$POS_PARMS $1"
+      shift
+      ;;
+  esac
+done
+
+# Convert namespace and tenant to all lower-case
+tenantNS=$(echo "$tenantNS"| tr '[:upper:]' '[:lower:]')
+tenant=$(echo "$tenant"| tr '[:upper:]' '[:lower:]')
+
+# Check for parameters - set with cluster or tenant
+if [ -z "$tenantNS" ] && [ -z "$tenant" ]; then
+  cluster="true"
+  namespace=$MON_NS
+  v4mSecret="v4m-grafana"
+elif [ -n "$tenantNS" ] && [ -n "$tenant" ]; then
+  cluster="false"
+  namespace=$tenantNS
+  v4mSecret="v4m-grafana-${tenant}"
+else
+  log_error "Both a [NAMESPACE] and a [TENANT] are required in order to change the Grafana admin password.";
+  exit 1
+fi
+
+# Check and make sure that a password was provided.
+if [ -z $password ]; then
+  log_error "A value for parameter [PASSWORD] has not been provided"
+  exit 1
+fi 
+
+if [ "$cluster" == "true" ]; then
+  grafanaPod="$(kubectl get pods -n $MON_NS -l app.kubernetes.io/name=grafana --template='{{range .items}}{{.metadata.name}}{{end}}')"
+else
+  grafanaPod="$(kubectl get pods -n $tenantNS -l app.kubernetes.io/name=grafana -l app.kubernetes.io/instance=v4m-grafana-${tenant} --template='{{range .items}}{{.metadata.name}}{{end}}')"
+fi
+
+# Error out if a Grafana pod has not been found
+if [ -z $grafanaPod ]; then
+  log_error "Unable to update Grafana password."
+  log_error "No Grafana pods were available to update"
+  exit 1
+fi
+
+# Changes the admin password using the Grafana CLI
+log_info "Updating the admin password using Grafana CLI"
+kubectl exec -n $namespace $grafanaPod -c grafana -- bin/grafana-cli admin reset-admin-password $password
+
+# Exit out of the script if the Grafana CLI call fails
+if (( $? == 1 )); then
+  log_error "An error occurred when updating the password"
+  exit 1
+fi
+
+# Patch new password in Kubernetes
+encryptedPassword="$(echo -n "$password" | base64)"
+log_info "Updating Grafana secret with the new password"
+kubectl -n $namespace patch secret $v4mSecret --type='json' -p="[{'op' : 'replace' ,'path' : '/data/admin-password' ,'value' : '$encryptedPassword'}]"
+
+# Restart Grafana pods and wait for them to restart
+log_info "Restarting the Grafana pods"
+kubectl delete pods -n $namespace -l "app.kubernetes.io/instance=v4m-prometheus-operator" -l "app.kubernetes.io/name=grafana"
+kubectl -n $namespace wait pods --selector "app.kubernetes.io/instance=v4m-prometheus-operator","app.kubernetes.io/name=grafana" --for condition=Ready --timeout=2m
```

Comment:
Why the wait?  Is this just to make sure the pod is restarted before the user sees the message reporting success?  What happens if the wait fails?  Should you be checking for that and providing an ERROR or WARNING message?

Reply:
I have made a habit of doing a wait when I restart pods just to make sure that the pods are in a "usable" state by the time the script is done.  I included a similar wait in the create_logging_datasource script, as well.